cloud-storage/cloud_storage.c

```diff
@@ -0,0 +1,888 @@
+   {
+      remote_file = (cloud_storage_item_t *)calloc(1, sizeof(cloud_storage_item_t));
+      remote_file->item_type = CLOUD_STORAGE_FILE;
+      remote_file->name = strdup(find_last_slash(local_file) + 1);
+      remote_file->next = NULL;
+      new_file = true;
+   }
+
+   if (_provider->upload_file(
+      remote_folder,
+      remote_file,
+      local_file))
+   {
+      if (!remote_file->type_data.file.hash_value) {
+         switch (remote_file->type_data.file.hash_type)
+         {
+            case CLOUD_STORAGE_HASH_MD5:
+               remote_file->type_data.file.hash_value = (char *)cloud_storage_get_md5_hash(local_file);
+               break;
+         }
+      }
+
+      remote_file->last_sync_time = time(NULL);
+
+      if (new_file)
+      {
+         remote_file->next = remote_folder->type_data.folder.children;
+         remote_folder->type_data.folder.children = remote_file;
+      }
+   }
+}
+
+static void _update_existing_item(cloud_storage_item_t *item_to_update, cloud_storage_item_t *other)
+{
+   if (item_to_update->name)
+   {
+      free(item_to_update->name);
+      item_to_update->name = NULL;
+   }
+
+   if (other->name)
+   {
+      item_to_update->name = strdup(other->name);
+   }
+
+   if (item_to_update->item_type == CLOUD_STORAGE_FILE)
+   {
+      if (item_to_update->type_data.file.download_url)
+      {
+         free(item_to_update->type_data.file.download_url);
+         item_to_update->type_data.file.download_url = NULL;
+      }
+
+      if (item_to_update->type_data.file.hash_value)
+      {
+         free(item_to_update->type_data.file.hash_value);
+         item_to_update->type_data.file.hash_value = NULL;
+      }
+   }
+
+   item_to_update->item_type = other->item_type;
+   if (other->item_type == CLOUD_STORAGE_FILE)
+   {
+      if (other->type_data.file.download_url)
+      {
+         item_to_update->type_data.file.download_url = strdup(other->type_data.file.download_url);
+      }
+
+      item_to_update->type_data.file.hash_type = other->type_data.file.hash_type;
+
+      if (other->type_data.file.hash_value)
+      {
+         item_to_update->type_data.file.hash_value = strdup(other->type_data.file.hash_value);
+      }
+   } else
+   {
+      item_to_update->type_data.folder.children = other->type_data.folder.children;
+   }
+}
+
+static void _sync_files_upload(folder_type_t folder_type)
+{
+   global_t *global;
+   cloud_storage_item_t *remote_folder;
+   char *raw_dir;
+   char *dir_name;
+   struct RDIR *dir;
+   char *filename;
+
+   global = global_get_ptr();
+   if (!global)
+   {
+      return;
+   }
+
+   switch (folder_type)
+   {
+      case CLOUD_STORAGE_GAME_STATES:
+         remote_folder = _game_states;
+         raw_dir = global->name.savestate;
+         break;
+      default:
+         return;
+   }
+
+   if (!remote_folder || strlen(raw_dir) == 0)
+   {
+      return;
+   }
+
+   dir_name = strdup(raw_dir);
+   if (strrchr(dir_name, '.') > strrchr(dir_name, '/'))
```

Comment:
The code has literally 0 comments. This is one of those places where one comment would save me some time :)

Reply:
I am not clear on what you are looking for here? I realize the code is far from acceptable quality, but there are now many comments. There is a comment two lines up and five lines down.

Reply:
Yeah previously there weren't any comments at all, that's why I added the comment. Thanks!
However it is still not clear to me why you do this conditionally. Why do you need to *sometimes* call basedir and some times not?
What kind of file path do you expect to receive here? I take you receive a file (xx.yyy) and you get the basedir to get the directory it's stored in right? Why not checking whether the path is a file, and if it is, take the base dir? Would it achieve the same? I just have the feeling (but I can be totally wrong! not pointing any fingers!) that this looks like a hack.

Reply:
Save state can have values such as
- /home/user/.local/retroarch/states/.state
- /home/user/.local/retroarch/states

In the first case, it is specifying a directory with an extension. In the second case it only specifies a directory. If it is the first case, I need to remove the extension. In the second case, I need to preserve the path as it is.

Far better would be to have separate settings for the save state directory and extension. I'm fine making that change, but it would be a separate PR and have concerns around backwards compatibility.